063-prometheus-metrics-reporter.md

```diff
@@ -0,0 +1,207 @@
+
+## Affected/not affected projects
+
+The reporter will be a new project, `strimzi-metrics-reporter`. It will be used by Strimzi components but should also be usable without Strimzi, for example in client side applications.
+
+### strimzi-kafka-operator
+
+Reporters are not usable with ZooKeeper so Strimzi should keep support for jmx_exporter while it also supports ZooKeeper. The transition would happen in 3 stages:
+
+1. The new reporters are added as an additional options to the Strimzi custom resource, but we keep supporting the JMX Exporter and use it in our examples as the main thing (with possibly some new examples of CRs and dashboards prepared as well, but not as the main thing)
+
+2. We switch to the new reporters as the main example. We keep supporting JMX Exporter but deprecate it.
+
+3. Later, once ZooKeeper is not supported anymore, we might decide to drop the JMX Exporter support completely.
+
+### strimzi-kafka-oauth
+
+The OAuth plugin exposes the `strimzi.oauth.metric.reporters` configuration to let users specify a metrics reporter. Today the OAuth plugin automatically uses `org.apache.kafka.common.metrics.JmxReporter` if that configuration is not set. To switch to the new reporter, users should set `strimzi.oauth.metric.reporters` to `KafkaPrometheusMetricsReporter`.
+
+Note that directly instantiating metric reporters in the plugin is a workaround as Kafka currently does not provide a way for plugins to register metrics. [KIP-877](https://cwiki.apache.org/confluence/display/KAFKA/KIP-877%3A+Mechanism+for+plugins+and+connectors+to+register+metrics), currently in discussion, aims at solving this issue. Once this KIP is accepted, we should update the plugin to use this mechanism.
+
+### strimzi-kafka-bridge
+
+To function, the bridge instantiates Kafka clients. Today it has its own custom mechanism to retrieve metrics from the Kafka clients via JMX and expose them to Prometheus.
+
+We can update the bridge to use metric reporters to retrieve metrics from Kafka clients and still keep it's own mechanism to expose them to Prometheus (as it's also exposing its own metrics). To switch behavior, I propose introducing a new configuration to the bridge `metricsMode`/`KAFKA_BRIDGE_METRICS_MODE` which would initially default to `jmx` to keep the current behavior. If set to `reporter`, the bridge would set the `metric.reporters` configuration on all Kafka clients it starts to `KafkaPrometheusMetricsReporter` and retrieve their metrics via the Prometheus metrics registry. It should also set `prometheus.metrics.reporter.listener` to `disabled` so `KafkaPrometheusMetricsReporter` instances don't start their own HTTP endpoint.
```

Comment:
Now I understand why, maybe, you needed disabled here. Isn't it enough just leaving out the parameter?